### DISCUSSION TITLE
Problem with Instantiating ExampleRocketSystem

### DISCUSSION THREAD

**Discussion by HaogeL** [chipsalliance/rocket-chip#3136] [Q&A] [answered]
Dear community, 

I encounter the following error when I tried to instantiate [ExampleRocketSystem ](https://github.com/chipsalliance/rocket-chip/blob/master/src/main/scala/system/ExampleRocketSystem.scala)in my own top class module.

> Exception in thread "main" firrtl.passes.CheckFlows$WrongFlow:  @[MyRocketchip.scala 57:49]: [module MyRocketchip]  Expression LazyRocketchipOri.l2_frontend_bus_axi4.0 is used as a SourceFlow but can only be used as a SinkFlow.

Basically, I created a class(MyRocketchip), declaring the same AXI buses (mmio, mem, and l2_frontbus) as [ExampleRocketSystem ](https://github.com/chipsalliance/rocket-chip/blob/master/src/main/scala/system/ExampleRocketSystem.scala), and tried to make the connections between the top-level bus and its corresponding bus in [ExampleRocketSystem ](https://github.com/chipsalliance/rocket-chip/blob/master/src/main/scala/system/ExampleRocketSystem.scala):
```scala
class MyRocketchip(implicit val p: Parameters) extends Module {
  val LazyRocketchipOri = LazyModule(new ExampleRocketSystem)
  val RocketchipOri= Module(LazyRocketchipOri.module)
  //tie-off l2_frontend_bus
  //LazyRocketchipOri.l2_frontend_bus_axi4.foreach(_.tieoff)
  val io = IO(new Bundle {
    val interrupts = Input(UInt(p(NExtTopInterrupts).W))
    val mem_axi4 = RocketchipOri.outer.mem_axi4.head.cloneType
    val mmio_axi4 = RocketchipOri.outer.mmio_axi4.head.cloneType
    val frontbus_axi4 = RocketchipOri.outer.l2_frontend_bus_axi4.head.cloneType
    val jtag = Flipped(new JTAGIO())
  })
....
....
  io.mem_axi4 <> RocketchipOri.outer.mem_axi4.head
  io.mmio_axi4 <> RocketchipOri.outer.mmio_axi4.head
  RocketchipOri.interrupts := io.interrupts
  RocketchipOri.outer.l2_frontend_bus_axi4.head <> io.frontbus_axi4
....
....
```
The connections/assignments of mem_axi4, mmio_axi4, and interrupts **work fine**. However, the error points to the last line in above code snippet (i.e. _RocketchipOri.outer.l2_frontend_bus_axi4.head <> io.frontbus_axi4_)

I appreciate if someone can explain why is that?  How can I propagate the AXI4 buses from ExampleRocketSystem to an upper level module?

Thanks in advance! 👍
Best regards,
HG

**Comment by michael-etzkorn** [comment]
I would consider turning your rocket-chip module into a LazyModule and connecting with diplomatic connections to the `ExampleRocketSystem`.

**Reply by HaogeL** [reply]
Thanks for your reply.
Yes, I thought about that, but for me, it is much more difficult to understand how to work with Diplomacy than just writing Chisel code.
I am in the middle of doing that anyway.

BTW, do you know is there any material or lab about Diplomacy? The [adder_tutorial](https://github.com/chipsalliance/rocket-chip/blob/master/docs/src/diplomacy/adder_tutorial.md) is abstruse😒

Best regards,
HG

**Comment by HaogeL** [comment, marked as answer]
Solved by applying **Flipped** to **frontbus_axi4**, although I don't understand why...
```scala
   val frontbus_axi4 = Flipped(RocketchipOri.outer.l2_frontend_bus_axi4.head.cloneType)
```
